src/Ganeti/HTools/Backend/IAlloc.hs

```diff
@@ -233,8 +234,19 @@ parseData now body = do
       dsrd_loc_tags = extractDesiredLocations ctags
       updateTags = updateExclTags ex_tags .
                    updateDesiredLocationTags dsrd_loc_tags
-  cdata1 <- mergeData [] [] [] [] now (ClusterData gl nl il ctags defIPolicy)
-  let (msgs, fix_nl) = checkData (cdNodes cdata1) (cdInstances cdata1)
+  -- hypervisor
+  enabled_hypervisors <- extrObj "enabled_hypervisors"
+  -- This is ugly, unfortunately the  IAllocator "API" (which is a CLI not an
+  -- API to be precise) is different from other htools backends that receive
+  -- the 'default_hypervisor' field. Normally there is no reason to have
+  -- more than one hypervisor so in practice this should be fine. Ganeti 2.17
+  -- exports the hv_stat per node field with the missing information.
+  -- TODO: Remove, once the caller can get the correct nodes size on KVM.
+  let nl2 = case enabled_hypervisors of
```

Comment:
I'd prefer if this was some sort of "plugin" or middleware instead of being hardcoded within the function: e.g.: `parseData :: ClockTime -> String -> (JSRecord -> a? -> b?) -> Result ...`. That plugin would then be applied here. However, the return types at this function (Result [String], Request]) are fairly broad, so it is hard to figure out what a? and b? are.

Do you have any ideas?



 The return types are not helping here, however